### ISSUE TITLE
Considerations for native compilation (gccemacs)

### ISSUE THREAD

**Issue by A6GibKm** [djcb/mu#1782]
## Expected or desired behavior

According to the issue described in #370, the `native-comp` branch of emacs ([gccemacs](http://akrl.sdf.org/gccemacs.html)) requires natively compiled elisp files to achieve enlightenment. In the best case scenario they are precompiled and readily available at runtime, if not they are built (if having permissions to the location of the `.elc` files) when needed in an async fashion.

## Actual behavior

The function `mu4e` will produce warnings as the lisp files are located in `/usr/share/emacs/...`, which needs root permissions.

## Steps to reproduce

run `mu4e` on the `native-comp` branch of emacs 28.0.50.

**EDIT:** It seems this only happens when the user enables
```lisp
(setq comp-deferred-compilation t)
```

## Versions of mu, mu4e/emacs, operating system etc.

* mu 1.4.13
* emacs 28.0.25 `native-comp` branch built with `configure --with-nativecomp`

## Any other detail

There are two possible ways to solve this as fast as I can think:

* Distribute mu4e in a portable fashion, see #370. This might prove problematic as `mu` is distributed in various Linux distributions with different versions and `mu4e` won't work if its version does not coincide with `mu`'s.

* Check in the `configure` step for the `native-comp` branch, maybe using `(fboundp 'native-compile)`, and run some sort of `emacs -Q -batch -L . -f batch-native-compile *.el` on the `make` or `make install` phase after `batch-byte-compile`ing the sources.

**Comment by djcb** [comment]
I'll take a look once the native-comp branch is merged; but I guess it would probably be some updated compilation step.

**Comment by andersjohansson** [comment]
This shouldn't be a problem any longer after the latest updates to the native-comp branch. Basically, native compiled .eln files will now be put in a cache directory, either a system directory that is written during installation of emacs, or a user writable one where all files compiled when running emacs is put.

Perhaps there could be a way of triggering native compilation of mu4e elisp already when invoking make install, so as to put the compiled files in the system cache directory when mu+mu4e is installed systemwide. Otherwise the mu4e elisp will be native compiled when loaded and put in the user cache directory, and everything should work.

See:
https://akrl.sdf.org/gccemacs.html#org4b11ea1
https://lists.gnu.org/archive/html/emacs-devel/2020-08/msg00535.html

**Comment by A6GibKm** [comment]
Yes, this won't be an issue any more, Still, it might be considered (later, when things are more stabilized) to natively compile files in the make/make install step.

**Comment by djcb** [comment]
Thanks! So, I guess I'll close this for now, once the stuff is merged and mu4e doesn't work with somehow, please file a new ticket.

**Comment by jian-lin** [comment]
Could we re-open this? @djcb 

Native compilation is a feature of emacs 28. And enabling native compilation in the build process of mu is good for distributions like NixOS to enable AOT native compilation.

**Comment by djcb** [comment]
@jian-lin: it seems to work fine as-is. AOT has low-prio for me at least, but If someone makes a really simple PR for this, we can take a look. Is there some other package that does this?

**Comment by jian-lin** [comment]
https://github.com/djcb/mu/pull/2343
